### DISCUSSION TITLE
OIDC State cookie not found

### DISCUSSION THREAD

**Discussion by jamesacris** [OpenIDC/mod_auth_openidc#1050] [Q&A]
Hi. I'm attempting to use mod_auth_openidc but I'm seeing 'no state cookie found' errors. After login, the state cookie goes missing, and mod_auth_openidc redirects me back to the homepage, which redirects me back to the OIDC provider and a loop ensues, broken by a 'too many redirects' error in the browser.

<details>
    <summary> Snapshot of apache debug logs, just before the error (client IP & hostname redacted):</summary>

```
...
[auth_openidc:debug] [pid 409:tid 140427999024896] src/authz.c(199): [client <ip>] oidc_authz_match_claim: evaluating key "groups"
[auth_openidc:debug] [pid 409:tid 140427999024896] src/authz.c(451): [client <ip>] oidc_authz_worker24: require claim/expr 'groups~misp*' matched
[authz_core:debug] [pid 409:tid 140427999024896] mod_authz_core.c(820): [client <ip>] AH01626: authorization result of Require claim "groups~misp*": granted
[authz_core:debug] [pid 409:tid 140427999024896] mod_authz_core.c(820): [client <ip>] AH01626: authorization result of <RequireAny>: granted
...
[auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(4045): [client <ip>] oidc_authz_checker: enter: require_args=""groups~misp*""
[authz_core:debug] [pid 409:tid 140427378292480] mod_authz_core.c(820): [client <ip>] AH01626: authorization result of Require claim "groups~misp*": denied (no authenticated user yet)
[authz_core:debug] [pid 409:tid 140427378292480] mod_authz_core.c(820): [client <ip>] AH01626: authorization result of <RequireAny>: denied (no authenticated user yet)
[auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(3928): [client <ip>] oidc_check_user_id: incoming request: "/oauth2callback?code=HZBOnil3vLUjUuGHu0qbPN&state=YizH13F9Vi_SFL98d7mq_w", ap_is_initial_req(r)=1
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(2521): [client <ip>] oidc_util_hdr_in_get: Cookie=mod_auth_openidc_session=8c67c0f5-0377-420c-a86f-a95de936e718
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1226): [client <ip>] oidc_util_get_cookie: returning "mod_auth_openidc_session" = "8c67c0f5-0377-420c-a86f-a95de936e718"
[auth_openidc:debug] [pid 409:tid 140427378292480] src/cache/common.c(318): [client <ip>] oidc_cache_get: enter: 8c67c0f5-0377-420c-a86f-a95de936e718 (section=s, decrypt=0, type=shm)
[auth_openidc:debug] [pid 409:tid 140427378292480] src/cache/common.c(350): [client <ip>] oidc_cache_get: cache hit: return 4203 bytes from shm cache backend for key 8c67c0f5-0377-420c-a86f-a95de936e718
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1388): [client <ip>] oidc_util_request_matches_url: comparing "/oauth2callback"=="/oauth2callback"
[auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(2104): [client <ip>] oidc_handle_redirect_authorization_response: enter
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1712): [client <ip>] oidc_util_read_form_encoded_params: read: code=HZBOnil3vLUjUuGHu0qbPN
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1712): [client <ip>] oidc_util_read_form_encoded_params: read: state=YizH13F9Vi_SFL98d7mq_w
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1717): [client <ip>] oidc_util_read_form_encoded_params: parsed: 56 bytes into 2 elements
[auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(1922): [client <ip>] oidc_handle_authorization_response: enter, response_mode=query
[auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(1547): [client <ip>] oidc_authorization_response_match_state: enter (state=YizH13F9Vi_SFL98d7mq_w)
[auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(649): [client <ip>] oidc_restore_proto_state: enter
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(2521): [client <ip>] oidc_util_hdr_in_get: Cookie=mod_auth_openidc_session=8c67c0f5-0377-420c-a86f-a95de936e718
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(2521): [client <ip>] oidc_util_hdr_in_get: Cookie=mod_auth_openidc_session=8c67c0f5-0377-420c-a86f-a95de936e718
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1226): [client <ip>] oidc_util_get_cookie: returning "mod_auth_openidc_state_YizH13F9Vi_SFL98d7mq_w" = <null>
[auth_openidc:error] [pid 409:tid 140427378292480] [client <ip>] oidc_restore_proto_state: no "mod_auth_openidc_state_YizH13F9Vi_SFL98d7mq_w" state cookie found: check domain and samesite cookie settings
[auth_openidc:error] [pid 409:tid 140427378292480] [client <ip>] oidc_authorization_response_match_state: unable to restore state
[auth_openidc:warn] [pid 409:tid 140427378292480] [client <ip>] oidc_handle_authorization_response: invalid authorization response state; a default SSO URL is set, sending the user there: http://<server_hostname>:8080
[auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(2577): [client <ip>] oidc_util_hdr_table_set: Location: http://<server_hostname>:8080
``` 
</details>

<details>
    <summary> Relevant config file (hostname replaced with {server_hostname}, OIDC provider replaced with {provider}):</summary>

```html
ServerTokens Prod

ServerName {server_hostname}:8080

# Include request ID header in accesss log
LogFormat "%h %{X-Request-Id}i %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" combined
LogLevel debug

<VirtualHost *:80>
    DocumentRoot /var/www/MISP/app/webroot

    ErrorDocument 401 /401.html
    ErrorDocument 403 /401.html
    ErrorDocument 500 /500.html
    ErrorDocument 503 /503.html
    ErrorDocument 504 /504.html
    Alias /401.html /var/www/html/401.shtml
    Alias /500.html /var/www/html/500.shtml
    Alias /503.html /var/www/html/503.shtml
    Alias /504.html /var/www/html/504.shtml

    <Directory /var/www/html/>
      Options +Includes
    </Directory>

    # Allow access to error page without authentication
    <LocationMatch "/(401|500).html">
        Satisfy any
    </LocationMatch>

    # Allow to access fpm-status just from localhost
    <LocationMatch "/fpm-status">
        Require local
        ProxyPass "unix:/run/php-fpm/www.sock|fcgi://127.0.0.1:9000"
    </LocationMatch>

    SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
    DirectoryIndex /index.php index.php
    <FilesMatch \.php$>
        SetHandler "proxy:unix:/run/php-fpm/www.sock|fcgi://127.0.0.1:9000"
    </FilesMatch>

    RewriteEngine On
    # Check if authkey is valid before we let apache to touch PHP
    RewriteMap authkeys "prg:/var/www/MISP/app/Console/cake user authkey_valid" apache:apache

    <Directory /var/www/MISP/app/webroot>
        Options -Indexes
        Require all granted

        # If request contains Authorization header and it is not request for fpm-status, check if authorizatzion key is valid.
        # This adds another level of protection.
        RewriteCond %{HTTP:Authorization} "^[a-zA-Z0-9]{40}$"
        RewriteCond %{REQUEST_URI} !^/fpm-status
        RewriteCond ${authkeys:%{HTTP:Authorization}|0} "!=1"
        # If authkey is not valid, return forbidden error
        RewriteRule .* - [F,L]

        # Standard MISP rules that will allow processing requests by PHP if it is not directory or file
        RewriteCond %{REQUEST_FILENAME} !-d
        RewriteCond %{REQUEST_FILENAME} !-f
        RewriteRule ^(.*)$ index.php?/$1 [QSA,L]
    </Directory>

    OIDCProviderMetadataURL https://{provider}/.well-known/openid-configuration
    OIDCRedirectURI http://{server_hostname}:8080/oauth2callback
    OIDCCryptoPassphrase {secret}
    OIDCClientID {secret}
    OIDCClientSecret {secret}
    OIDCDefaultURL http://{server_hostname}:8080
    OIDCCookieSameSite On
    OIDCProviderTokenEndpointAuth client_secret_jwt
    OIDCPKCEMethod S256
    OIDCScope "openid email profile"

    OIDCHTMLErrorTemplate /var/www/html/oidc.html

    # Allow access if header contains Authorization header and value in MISP format
    <If "-T req('Authorization') && req('Authorization') =~ /^[a-zA-Z0-9]{40}$/">
        Require all granted
        AuthType None
    </If>
    <Else>
        AuthType openid-connect
        Require claim "groups~misp*"
    </Else>

    TimeOut 310
    ServerSignature Off

    # Set request ID if not set from reverse proxy
    RequestHeader setifempty X-Request-Id %{UNIQUE_ID}e

    Header always set X-Content-Type-Options nosniff
    Header always set X-Frame-Options SAMEORIGIN
    Header always set Referrer-Policy same-origin

    <Location "/webfonts/">
        # Cache for one year
        Header always set Cache-Control "max-age=31536000; immutable"
    </Location>

    # Enable brotli ouput compression
    AddOutputFilterByType BROTLI_COMPRESS text/html text/plain text/xml text/css text/javascript application/x-javascript application/javascript application/json application/x-font-ttf image/svg+xml
</VirtualHost>
``` 
</details>

Some things that seem weird to me:
1. There is a rule in the config to check that the groups claim of the JWT contains the string misp. It looks like it validates it, but then further down it denies, citing 'no authenticated user'
2. oidc_util_get_cookie returns null right before the error message that the cookie cannot be found. Why would the cookie be missing?

From the wiki on [state cookies](https://github.com/OpenIDC/mod_auth_openidc/wiki/Cookies#state-cookie):

> [The state cookie] is deleted when the user returns to the Apache server with an authentication response (indicating either success or failure).

Two possibilities in my mind:
1. The multiple checks against the groups claim is meaning the cookie was deleted but is then needed again for some reason
2. Redirects are not working properly (though I have checked that all the redirect urls match in the OIDC provider) and the page that is being landed on does not have the cookie that's being looked for.

I'm struggling to identify the cause of the error, any help would be greatly appreciated 🙂

**Comment by zandbelt** [comment]
Firstly, you should not run over plain HTTP...
Then there seems to be a mix of running on port 8080 (ServerName, OIDCRedirectURI) and port 80 (VirtualHost), which one should it be?
Finally, what version of the module are you on and what are you trying to achieve with `OIDCCookieSameSite On`?

**Reply by jamesacris** [reply]
Hi @zandbelt, thanks for the prompt response. I probably should have been clearer in my initial message, here are some more details about the deployment:

I'm using the [NUKIB MISP docker image](https://github.com/NUKIB/misp) so the config is similar to what is used there. Their out-of-the-box Apache config can be found in [misp.conf](https://github.com/NUKIB/misp/blob/main/misp.conf). Since this is a containerised deployment the server is run on port 80 in the container which maps to port 8080 on the host (see [the docker compose file](https://github.com/NUKIB/misp/blob/812c30f3608f477bc72a25679ed98fb2130b5c4a/docker-compose.yml#L66) as an example). Regarding HTTP the intention is to deploy reverse proxy in front of MISP to handle HTTPS connections for production.

mod_auth_openidc version 2.4.9.4

The `OIDCCookieSameSite On` setting is one that was set in the upstream config - I believe it's to ensure the original login request originated from the same site, but I'm not exactly sure. I have tried turning it off to see if that helps, unfortunately it doesn't.

**Comment by zandbelt** [comment]
it looks like a mismatch in hostname used for accessing the application and the redirect URI; there should also be a warning about that in the logs; what is the URL that you're using in the browser to access the protected application?

**Reply by jamesacris** [reply]
Ok, I'll check to see if I can find any warnings about that in the logs.

The url I'm going to in the browser is `http://host-172-16-100-82.nubes.stfc.ac.uk:8080/`
It matches the first part of the redirect uri in the config which is:
`OIDCRedirectURI http://host-172-16-100-82.nubes.stfc.ac.uk:8080/oauth2callback`
As well as the redirect uri in the OIDC provider:
`	"redirectUris": [
		"http://host-172-16-100-82.nubes.stfc.ac.uk:8080/oauth2callback"
	],`

**Reply by zandbelt** [reply]
ok, that looks good to me

**Comment by zandbelt** [comment]
upon hitting the protected website there's a redirect with an authentication request returned, including a `Set-Cookie` header; can you paste that header here?

**Reply by jamesacris** [reply]
Here is an example from a login attempt just performed now... the cookie expired in 1970... I understood that was sometimes done by sites as a way to delete the cookie. Is that the case here? Very strange.

```
Request URL: http://host-172-16-100-82.nubes.stfc.ac.uk:8080/oauth2callback?code=EFeDC9Sk8kPJA7p7M7RzrM&state=_lnHLIW9rMCb5YYQBpfGRw_aIwI
...
Set-Cookie: mod_auth_openidc_state__lnHLIW9rMCb5YYQBpfGRw_aIwI=; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT; HttpOnly
Set-Cookie: mod_auth_openidc_session=b5f50e0b-f9d1-4af6-a0b1-bf585743e2bf; Path=/; HttpOnly; SameSite=Lax
```

**Reply by zandbelt** [reply]
this all looks good, a session is created and the state cookie is deleted; perhaps there's an application error afterwards and the browser (some browsers do...) re-tries the request?

**Reply by jamesacris** [reply]
Interesting, ok. I'll see if I can find any relevant application logs. Thanks for your help so far.

**Reply by jamesacris** [reply]
All I can see in the application logs is the repeated redirects from the loop that ensues after login. I can't find a record of what actually causes the error.
```
- ZEaGDGQ6nS1TYDgYnxFjvQAAANU -  24/Apr/2023:13:37:16 +0000 "GET /index.php" 302 28.665 2048 139.54%
- ZEaGDGQ6nS1TYDgYnxFjvgAAANY -  24/Apr/2023:13:37:16 +0000 "GET /index.php?/users/login" 302 30.526 2048 98.28%
- ZEaGDGQ6nS1TYDgYnxFjwAAAANg -  24/Apr/2023:13:37:16 +0000 "GET /index.php" 302 28.589 2048 104.94%
- ZEaGDGQ6nS1TYDgYnxFjwQAAAMA -  24/Apr/2023:13:37:16 +0000 "GET /index.php?/users/login" 302 31.364 2048 95.65%
- ZEaGDGQ6nS1TYDgYnxFjwwAAAMI -  24/Apr/2023:13:37:16 +0000 "GET /index.php" 302 27.609 2048 108.66%
...
```

All the information I have on what went wrong looks to be here, in the Apache debug logs.
```
Apr 25 11:13:11 9cf6fd7d12cc apache-error [Tue Apr 25 11:13:11.304667 2023] [auth_openidc:debug] [pid 136:tid 139788787095296] src/util.c(1226): [client 130.246.161.157:60215] oidc_util_get_cookie: returning "mod_auth_openidc_state_iZljvVxrokMzbkuvSpv9Ag" = <null>
Apr 25 11:13:11 9cf6fd7d12cc apache-error [Tue Apr 25 11:13:11.304682 2023] [auth_openidc:error] [pid 136:tid 139788787095296] [client 130.246.161.157:60215] oidc_restore_proto_state: no "mod_auth_openidc_state_iZljvVxrokMzbkuvSpv9Ag" state cookie found: check domain and samesite cookie settings
Apr 25 11:13:11 9cf6fd7d12cc apache-error [Tue Apr 25 11:13:11.304714 2023] [auth_openidc:error] [pid 136:tid 139788787095296] [client 130.246.161.157:60215] oidc_authorization_response_match_state: unable to restore state
Apr 25 11:13:11 9cf6fd7d12cc apache-error [Tue Apr 25 11:13:11.304761 2023] [auth_openidc:warn] [pid 136:tid 139788787095296] [client 130.246.161.157:60215] oidc_handle_authorization_response: invalid authorization response state; a default SSO URL is set, sending the user there: http://host-172-16-100-82.nubes.stfc.ac.uk:8080
```
I assume the call to `oidc_util_get_cookie` shouldn't return `null` for the state cookie. Don't suppose you have any other suggestions?

**Reply by zandbelt** [reply]
I bet further up in the log you'll see entries like the successful one that you pasted before, for the same state cookie value, which would delete the cookie so it is not available the 2nd time around; you'd then have to find out why it comes around for the 2nd time

**Reply by jamesacris** [reply]
Ah yes, there are a few references to the same state cookie identifier (YizH13F9Vi_SFL98d7mq_w in this case). I have attached the logs below. It looks like:
1. `oidc_check_user_id` finds the state cookie in the incoming request
2. `oidc_util_read_form_encoded_params` reads it correctly
3. `oidc_authorization_response_match_state` allows the state cookie
4. `oidc_restore_proto_state` initially responds with 'enter'
5. `oidc_util_get_cookie` returns `<null>` instead of the actual state cookie value
6. `oidc_restore_proto_state` then complains that there is no state cookie.

So the question is, what causes the `oidc_util_get_cookie` function to return `<null>` even though the cookie has been correctly retrieved from the incoming request?
```
[Mon Apr 24 19:27:16.783012 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(3928): [client 130.246.160.95:58658] oidc_check_user_id: incoming request: "/oauth2callback?code=HZBOnil3vLUjUuGHu0qbPN&state=YizH13F9Vi_SFL98d7mq_w", ap_is_initial_req(r)=1

[Mon Apr 24 19:27:16.783032 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(2521): [client 130.246.160.95:58658] oidc_util_hdr_in_get: Cookie=mod_auth_openidc_session=8c67c0f5-0377-420c-a86f-a95de936e718

[Mon Apr 24 19:27:16.783049 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1226): [client 130.246.160.95:58658] oidc_util_get_cookie: returning "mod_auth_openidc_session" = "8c67c0f5-0377-420c-a86f-a95de936e718"

[Mon Apr 24 19:27:16.783068 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/cache/common.c(318): [client 130.246.160.95:58658] oidc_cache_get: enter: 8c67c0f5-0377-420c-a86f-a95de936e718 (section=s, decrypt=0, type=shm)

[Mon Apr 24 19:27:16.783191 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/cache/common.c(350): [client 130.246.160.95:58658] oidc_cache_get: cache hit: return 4203 bytes from shm cache backend for key 8c67c0f5-0377-420c-a86f-a95de936e718

[Mon Apr 24 19:27:16.783474 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1388): [client 130.246.160.95:58658] oidc_util_request_matches_url: comparing "/oauth2callback"=="/oauth2callback"

[Mon Apr 24 19:27:16.783506 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(2104): [client 130.246.160.95:58658] oidc_handle_redirect_authorization_response: enter

[Mon Apr 24 19:27:16.783569 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1712): [client 130.246.160.95:58658] oidc_util_read_form_encoded_params: read: code=HZBOnil3vLUjUuGHu0qbPN

[Mon Apr 24 19:27:16.783611 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1712): [client 130.246.160.95:58658] oidc_util_read_form_encoded_params: read: state=YizH13F9Vi_SFL98d7mq_w

[Mon Apr 24 19:27:16.783640 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1717): [client 130.246.160.95:58658] oidc_util_read_form_encoded_params: parsed: 56 bytes into 2 elements

[Mon Apr 24 19:27:16.783652 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(1922): [client 130.246.160.95:58658] oidc_handle_authorization_response: enter, response_mode=query

[Mon Apr 24 19:27:16.783671 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(1547): [client 130.246.160.95:58658] oidc_authorization_response_match_state: enter (state=YizH13F9Vi_SFL98d7mq_w)

[Mon Apr 24 19:27:16.783686 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/mod_auth_openidc.c(649): [client 130.246.160.95:58658] oidc_restore_proto_state: enter

[Mon Apr 24 19:27:16.783697 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(2521): [client 130.246.160.95:58658] oidc_util_hdr_in_get: Cookie=mod_auth_openidc_session=8c67c0f5-0377-420c-a86f-a95de936e718

[Mon Apr 24 19:27:16.783715 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(2521): [client 130.246.160.95:58658] oidc_util_hdr_in_get: Cookie=mod_auth_openidc_session=8c67c0f5-0377-420c-a86f-a95de936e718

[Mon Apr 24 19:27:16.783734 2023] [auth_openidc:debug] [pid 409:tid 140427378292480] src/util.c(1226): [client 130.246.160.95:58658] oidc_util_get_cookie: returning "mod_auth_openidc_state_YizH13F9Vi_SFL98d7mq_w" = <null>

[Mon Apr 24 19:27:16.783747 2023] [auth_openidc:error] [pid 409:tid 140427378292480] [client 130.246.160.95:58658] oidc_restore_proto_state: no "mod_auth_openidc_state_YizH13F9Vi_SFL98d7mq_w" state cookie found: check domain and samesite cookie settings

[Mon Apr 24 19:27:16.783762 2023] [auth_openidc:error] [pid 409:tid 140427378292480] [client 130.246.160.95:58658] oidc_authorization_response_match_state: unable to restore state
```

**Reply by jamesacris** [reply]
That particular state cookie identifier is not found again outside of these few log lines by the way, so I don't think it was used previously and therefore deleted then

**Reply by zandbelt** [reply]
you're conflating the "state" value in the authentication response and the associated state cookie that is apparently not found; yet there seems to be successful attempts as well as you showed before, probably a browser trace (developer mode) would show more about the different behaviours

**Reply by jamesacris** [reply]
Ah, I see - in which case looking at the browser trace - in the first request to the application url (before redirecting to SSO) I can see the state cookie:

![misp browser trace 3](https://user-images.githubusercontent.com/71633875/234296518-373ed33b-b841-44d3-a905-b7f3e24edae4.png)

Then, in the first request to the `/oauth2callback` url after login, there is the same state cookie, which then appears to be deleted in the response.

![misp browser trace 4](https://user-images.githubusercontent.com/71633875/234296947-44e916ba-8809-4a32-afad-ea719477762f.png)

![misp browser trace](https://user-images.githubusercontent.com/71633875/234292144-eb489edc-aed8-4ab4-b0df-c6c80c606fde.png)

I checked subsequent requests made by the browser to the same endpoint, after being redirected back to the homepage, OIDC provider, etc. and there was no state cookie, only a session cookie.

![misp browser trace 2](https://user-images.githubusercontent.com/71633875/234294178-78a716e5-26ba-411b-b525-92f46e244e20.png)

**Reply by zandbelt** [reply]
is is not supposed to be calling that endpoint again once it has established a session

**Reply by jamesacris** [reply]
Yes, it only does that after the state cookie can't be found, and therefore redirects me back to the homepage. That then redirects to the OIDC provider, which thinks I'm already authenticated, and so sends me back to that endpoint again. 
I still think the problem has something to do with why the `oidc_util_get_cookie` function is returning `<null>`. Is there a way to look more in depth at this that you know of?

**Reply by zandbelt** [reply]
you'll have to look into what happens after the session has been established, why is the web server not returning application content? my best guess is that the application somehow crashes/aborts on the provided authentication state which makes the browser issue another request with the same state that is invalid by then

**Reply by jamesacris** [reply]
Ok, the problem is quite possibly out of the mod_auth_openidc remit then. Thanks for your patience with this, I'll look into what could be going wrong with the application further.

**Comment by Abhikr1994** [comment]
Hey James,

I am also facing same issue.Can you tell whether you have got solution or not?

**Reply by jamesacris** [reply]
Sorry - I didn't figure it out. My approach is now to use a plugin for the underlying application itself to do OIDC rather than using Apache.
